rskj-core/src/main/java/org/ethereum/datasource/DataSourceWithCache.java

```diff
@@ -0,0 +1,155 @@
+        Objects.requireNonNull(key);
+        ByteArrayWrapper wrappedKey = ByteUtil.wrap(key);
+
+        if (uncommittedKeysToRemove.contains(wrappedKey)) {
+            return null;
+        }
+
+        if (committedCache.containsKey(wrappedKey)) {
+            return committedCache.get(wrappedKey);
+        }
+        if (uncommittedCache.containsKey(wrappedKey)) {
+            return uncommittedCache.get(wrappedKey);
+        }
+
+        byte[] value = base.get(key);
+        //null value, as expected, is allowed here to be stored in committedCache
+        committedCache.put(wrappedKey, value);
+        return value;
+    }
+
+    @Override
+    public byte[] put(byte[] key, byte[] value) {
+        Objects.requireNonNull(key);
+        Objects.requireNonNull(value);
+        ByteArrayWrapper wrappedKey = ByteUtil.wrap(key);
+
+        // here I could check for equal data or just move to the uncommittedCache.
+        byte[] priorValue = committedCache.get(wrappedKey);
+        if (priorValue != null && Arrays.equals(priorValue, value)) {
+            return value;
+        }
+
+        committedCache.remove(wrappedKey);
+        uncommittedCache.put(wrappedKey, value);
+        uncommittedKeysToRemove.remove(wrappedKey);
+        return value;
+    }
+
+    @Override
+    public void delete(byte[] key) {
+        Objects.requireNonNull(key);
+        ByteArrayWrapper wrappedKey = ByteUtil.wrap(key);
+        // fully delete this element
+        committedCache.remove(wrappedKey);
+        uncommittedCache.remove(wrappedKey);
+        uncommittedKeysToRemove.add(wrappedKey);
+    }
+
+    @Override
+    public Set<byte[]> keys() {
+        Stream<ByteArrayWrapper> baseKeys = base.keys().stream().map(ByteArrayWrapper::new);
+        Stream<ByteArrayWrapper> committedKeys = committedCache.entrySet().stream()
+                .filter(e -> e.getValue() != null)
+                .map(Map.Entry::getKey);
+        // there is no need to remove null values on uncommittedCache
+        Stream<ByteArrayWrapper> uncommittedKeys = uncommittedCache.keySet().stream();
+        Set<ByteArrayWrapper> knownKeys = Stream.concat(Stream.concat(baseKeys, committedKeys), uncommittedKeys)
+                .collect(Collectors.toSet());
+        knownKeys.removeAll(uncommittedKeysToRemove);
+
+        return knownKeys.stream()
+                .map(ByteArrayWrapper::getData)
+                .collect(Collectors.toSet());
+    }
+
+    @Override
+    public void updateBatch(Map<ByteArrayWrapper, byte[]> rows, Set<ByteArrayWrapper> keysToRemove) {
+        if (rows.containsKey(null) || rows.containsValue(null)) {
+            throw new IllegalArgumentException("Cannot update null values");
+        }
+
+        // don't move unchanged keys unnecessarily
+        Map<ByteArrayWrapper, byte[]> rowsToUpdate = new HashMap<>(rows);
+        rowsToUpdate.entrySet().removeIf(e -> Arrays.equals(e.getValue(), committedCache.get(e.getKey())));
```

Comment:
> if a large batch of keys are removed several times, updateBatch won't detect it, and every time the uncommittedKeysToRemove will be loaded by the same keys
I cannot see how it's less optimal to implement a mechanism to detect recurrent removal attempts than just add that collection to the `uncommittedKeysToRemove` set. It may be more readable to make the uncommitedCache support null semantics tough. I'll implement this as a separate commit so you can review it, and if it looks better I'll squash that solution before merge

Reply:
Because the "base" is assumed to be slower (for deletes, even if the element does not exists), and that's why the cache is needed.